faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -358,90 +352,82 @@ public void syncLastTxnTime(long newTxnTime) {
-    request.headers().set("X-FaunaDB-API-Version", API_VERSION);
+    URI requestUri = URI.create(mkUrl(path));
 
-    if(jvmDriver != null) {
-      request.headers().set(X_FAUNA_DRIVER, jvmDriver.toString());
+    // Encode all query parameters
+    for (Map.Entry<String,String> entry : params.entrySet()) {
+      requestUri = appendUri(requestUri, entry.getKey(), entry.getValue());
     }
 
-    // TODO: replace as follows when upgrade to Java 9 or higher:
-    // requestQueryTimeout.or(queryTimeout)
-    //   .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-    queryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
+
+    HttpRequest.BodyPublisher bodyPublisher = HttpRequest.BodyPublishers.noBody();
+    if (body.isPresent()) {
+      byte[] jsonBody = json.writeValueAsBytes(body.get());
+      bodyPublisher = HttpRequest.BodyPublishers.ofByteArray(jsonBody);
+    };
 
     // If a query timeout has been given for the current request,
     // override the one from the Connection if any
-    requestQueryTimeout
-      .ifPresent(timeout -> request.headers().set(X_QUERY_TIMEOUT, timeout.toMillis()));
-
-    long time = getLastTxnTime();
-    if (time > 0) {
-      request.headers().set("X-Last-Seen-Txn", Long.toString(time));
-    }
-
-    request.retain();
+    Optional<Duration> queryTimeout = requestQueryTimeout.or(() -> connectionQueryTimeout);
+
+    Optional<Long> lastTxnTime = (getLastTxnTime() > 0) ? Optional.of(getLastTxnTime()) : Optional.empty();
+
+    HttpRequest.Builder requestBuilder =
+      HttpRequest.newBuilder()
+        .uri(requestUri)
+        .version(HttpClient.Version.HTTP_1_1)
+        .timeout(Duration.ofMillis(DEFAULT_REQUEST_TIMEOUT_MS))
+        .method(httpMethod, bodyPublisher)
+        .headers(
+          "Authorization", authHeader,
+          "X-FaunaDB-API-Version", API_VERSION,
+          "User-agent", "Fauna JVM Http Client",
+          X_FAUNA_DRIVER, jvmDriver.toString(),
+          "Content-type", "application/json; charset=utf-8"
+        );
+
+    queryTimeout.ifPresent( timeout ->
+      requestBuilder.header(X_QUERY_TIMEOUT, String.valueOf(timeout.toMillis()))
+    );
+
+    lastTxnTime.ifPresent( time ->
+      requestBuilder.header("X-Last-Seen-Txn", Long.toString(time))
+    );
+
+    HttpRequest request = requestBuilder.build();
 
     client.sendRequest(request).whenCompleteAsync((response, throwable) -> {
-
       ctx.stop();
 
       if (throwable != null) {
         logFailure(request, throwable);
-        request.release();
-        if (response != null)
-          response.release();
         rv.completeExceptionally(throwable);
         return;
       }
 
-      String txnTimeHeader = response.headers().get("X-Txn-Time");
-      if (txnTimeHeader != null) {
-        syncLastTxnTime(Long.parseLong(txnTimeHeader));
-      }
+      Optional<String> txnTimeHeader = response.headers().firstValue("X-Txn-Time");
```

Comment:
FYI In my streaming PRs I have lower cased this header as this is the way core sends it.

Reply:
👍 I changed it here so as well to ease the review.